### PR TITLE
add native version for function tri

### PR DESCRIPTION
**Pull request by Rubtsowa**

### CODE REVIEW

File: tests/test_arraycreation.py [hunk 1]
```diff
@@ -110,3 +110,15 @@ def test_loadtxt(type):
...
+def test_triu_int():
+    expected = numpy.tri(3, 5, 2, dtype=int)
+    result = dpnp.tri(3, 5, 2, dtype=int)
+    numpy.testing.assert_array_equal(result, expected)
+
+
+def test_triu_float():
+    expected = numpy.tri(3, 5, -1)
+    result = dpnp.tri(3, 5, -1)
+    numpy.testing.assert_array_equal(result, expected)
```
Comment (samir-nasibli) [inline tests/test_arraycreation.py:124]: can't this be done as a parameterized test for different dtypes?

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 2]
```diff
@@ -72,6 +72,39 @@ void dpnp_full_c(void* array_in, void* result, const size_t size)
...
+        if (i + k + 1 > 0)
+        {
+            diag_idx_ = i + k + 1;
+        }
+
+        size_t diag_idx = diag_idx_;
+        if (M < diag_idx_)
+        {
+            diag_idx = M;
+        }
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:92]:
``` suggestion
        size_t diag_idx = (M < diag_idx_) ? M : diag_idx_;
```
See prev. comment.

**Review by samir-nasibli — COMMENTED** [review]
Again, no description provided.
It is needed some info about passed numpy original tests. It is important.

Also you don't testing corner cases on dpnp own tests, more test coverage is needed.

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 3]
```diff
@@ -72,6 +72,32 @@ void dpnp_full_c(void* array_in, void* result, const size_t size)
...
+void dpnp_tri_c(void* result1, const size_t N, const size_t M, const int k)
+{
+    _DataType* result = reinterpret_cast<_DataType*>(result1);
+
+    for (size_t i = 0; i < N; ++i)
+    {
+        int val = i + k + 1;
+        int diag_idx_ = (val > 0) ? val : 0;
+        int val_m = M;
+        size_t diag_idx = (val_m < diag_idx_) ? val_m : diag_idx_;
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:85]:
```suggestion
        size_t diag_idx_ = (val > 0) ? (size_t)val : 0;
        size_t diag_idx = (M < diag_idx_) ? M : diag_idx_;	
```

Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:85]: or
```suggestion
        size_t diag_idx = 0;
        if (val > 0) {
            diag_idx = (M < (size_t)val) ? M : (size_t)val;
        }
```

**Review by shssf — APPROVED** [review]

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 4]
```diff
@@ -116,14 +116,28 @@ void dpnp_ones_c(void* result, size_t size)
...
 template <typename _DataType>
-void dpnp_zeros_c(void* result, size_t size)
+void dpnp_tri_c(void* result1, const size_t N, const size_t M, const int k)
 {
-    _DataType* fill_value = reinterpret_cast<_DataType*>(dpnp_memory_alloc_c(sizeof(_DataType)));
-    fill_value[0] = 0;
+    _DataType* result = reinterpret_cast<_DataType*>(result1);
 
-    dpnp_initval_c<_DataType>(result, fill_value, size);
+    for (size_t i = 0; i < N; ++i)
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:123]: Don't we want to implement the loop with SYCL?

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 5]
```diff
@@ -353,13 +378,25 @@ void func_map_init_arraycreation(func_map_t& fmap)
...
+    fmap[DPNPFuncName::DPNP_FN_TRIL][eft_DBL][eft_DBL] = {eft_DBL, (void*)dpnp_tril_c<double>};
+
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_INT][eft_INT] = {eft_INT, (void*)dpnp_zeros_c<int>};
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_LNG][eft_LNG] = {eft_LNG, (void*)dpnp_zeros_c<long>};
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_FLT][eft_FLT] = {eft_FLT, (void*)dpnp_zeros_c<float>};
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_DBL][eft_DBL] = {eft_DBL, (void*)dpnp_zeros_c<double>};
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_BLN][eft_BLN] = {eft_BLN, (void*)dpnp_zeros_c<bool>};
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_C128][eft_C128] = {eft_C128, (void*)dpnp_ones_c<std::complex<double>>};
 
+<<<<<<< HEAD
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:398]: Please remove `git` symbols.

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 6]
```diff
@@ -121,35 +121,50 @@ void dpnp_ones_c(void* result, size_t size)
...
-            size_t ind = i * M + j;
-            result[ind] = 0;
+            int val = i + k + 1;
+            size_t diag_idx_ = (val > 0) ? (size_t)val : 0;
+            size_t diag_idx = (M < diag_idx_) ? M : diag_idx_;
+
+            for (size_t j = 0; j < diag_idx; ++j)
+            {
+                size_t ind = i * M + j;
+                result[ind] = 1;
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:154]: Just in case, is it possible here that different threads simultaneously write to the same place? I mean can we have the same `ind` for different threads?

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 7]
```diff
@@ -128,20 +128,45 @@ void dpnp_ones_like_c(void* result, size_t size)
...
-    _DataType* fill_value = reinterpret_cast<_DataType*>(dpnp_memory_alloc_c(sizeof(_DataType)));
-    fill_value[0] = 0;
+    cl::sycl::event event;
 
-    dpnp_initval_c<_DataType>(result, fill_value, size);
+    _DataType* result = reinterpret_cast<_DataType*>(result1);
 
-    dpnp_memory_free_c(fill_value);
-}
+    size_t idx = N* M;
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:140]: Will your algo work correct if N and M multiplication is zero?
Do you have any tests for this case?

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 8]
```diff
@@ -128,20 +128,45 @@ void dpnp_ones_like_c(void* result, size_t size)
 }
 
 template <typename _DataType>
-void dpnp_zeros_c(void* result, size_t size)
+class dpnp_tri_c_kernel;
+
+template <typename _DataType>
+void dpnp_tri_c(void* result1, const size_t N, const size_t M, const int k)
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:134]: No checks if `result1` is nullptr.

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp (hunk 6 above)
Comment (samir-nasibli) [reply dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:154]: `ind` takes threads `id`.

**Review by samir-nasibli — COMMENTED** [review]
Which numpy tests check your functionality? Are there any?
Provide, please, info about them (passed/disabled) on PR description.

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp (hunk 6 above)
Comment (shssf) [reply dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:159]: "I mean can we have the same ind for different threads?"
It is impossible by design. Each thread takes unique `id` by input parameter.

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 9]
```diff
@@ -128,20 +128,50 @@ void dpnp_ones_like_c(void* result, size_t size)
...
 
-template <typename _DataType>
-void dpnp_zeros_like_c(void* result, size_t size)
-{
-    dpnp_zeros_c<_DataType>(result, size);
+    size_t idx = N* M;
+    cl::sycl::range<1> gws(idx);
+    auto kernel_parallel_for_func = [=](cl::sycl::id<1> global_id) {
+        size_t ind = global_id[0];
+        {
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:149]: Do we really need this brackets here?

**Review by shssf — APPROVED** [review]

**Comment by shssf** [comment]
I see "TeamCity Build SAT" failed

**Comment by shssf** [comment]
`
DPNP_QUEUE_GPU=0 pytest "tests/test_arraycreation.py::test_tri[float64--5-0-0]"
`
passed

`DPNP_QUEUE_GPU=1 pytest "tests/test_arraycreation.py::test_tri[float64--5-0-0]"
`
failed with

```
DPNP: current queue is CPU: 0
DPNP: version: 0.5.1

collected 1 item

tests/test_arraycreation.py Fatal Python error: Aborted

Current thread 0x00007fbc7c671740 (most recent call first):
  File "/localdisk/work/sshalnov/dpnp/dpnp/dpnp_iface_arraycreation.py", line 1074 in tri
  File "/localdisk/work/sshalnov/dpnp/tests/test_arraycreation.py", line 155 in test_tri
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/python.py", line 184 in pytest_pyfunc_call
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/python.py", line 1627 in runtest
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/runner.py", line 163 in pytest_runtest_call
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/runner.py", line 256 in <lambda>
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/runner.py", line 310 in from_call
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/runner.py", line 255 in call_runtest_hook
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/runner.py", line 216 in call_and_report
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/runner.py", line 127 in runtestprotocol
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/runner.py", line 110 in pytest_runtest_protocol
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/main.py", line 338 in pytest_runtestloop
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/main.py", line 313 in _main
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/main.py", line 257 in wrap_session
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/main.py", line 306 in pytest_cmdline_main
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/localdisk/work/sshalnov/anaconda3/lib/python3.8/site-packages/_pytest/config/__init__.py", line 164 in main
  File "/localdisk/work/sshalnov/anaconda3/bin/pytest", line 11 in <module>

```

**Comment by shssf** [comment]
test case with
```
N=0
M=0
k=-5
dtype=<class 'numpy.float64'>
terminate called after throwing an instance of 'cl::sycl::runtime_error'
  what():  Native API failed. Native API returns: -30 (CL_INVALID_VALUE) -30 (CL_INVALID_VALUE)
Fatal Python error: Aborted

```

**Comment by samir-nasibli** [comment]
> test case with
> 
> ```
> N=0
> M=0
> k=-5
> dtype=<class 'numpy.float64'>
> terminate called after throwing an instance of 'cl::sycl::runtime_error'
>   what():  Native API failed. Native API returns: -30 (CL_INVALID_VALUE) -30 (CL_INVALID_VALUE)
> Fatal Python error: Aborted
> ```

I have asked this on https://github.com/IntelPython/dpnp/pull/623#discussion_r588456116 already.
Seems check for non-zero `N` an `M` is really needed.
